File: modules/meeterm-terminal/ios/AttachmentSession.swift
import Foundation

/**
 * Issue #28 one-attachment session model.
 *
 * JS sees only narrow result dictionaries; the staging/prepared file names are
 * kept here, and `snapshot` exposes display metadata plus the app-local
 * preview file URL.
 */
struct AttachmentTargetIdentity {
  let terminalId: String
  let paneId: String
  let workspaceId: String
  let backend: String
  let runtime: String
  let host: String
  let port: Int

  static func from(_ raw: [String: Any]?) -> AttachmentTargetIdentity? {
    guard let raw = raw,
          let terminalId = raw["terminalId"] as? String,
          let paneId = raw["paneId"] as? String,
          let workspaceId = raw["workspaceId"] as? String,
          let backend = raw["backend"] as? String,
          let runtime = raw["runtime"] as? String,
          let host = raw["host"] as? String,
          let port = raw["port"] as? Int ?? (raw["port"] as? NSNumber)?.intValue else {
      return nil
    }
    return AttachmentTargetIdentity(
      terminalId: terminalId,
      paneId: paneId,
      workspaceId: workspaceId,
      backend: backend,
      runtime: runtime,
      host: host,
      port: port
    )
  }
}

struct AttachmentPreparedImage {
  let fileName: String
  let format: AttachmentImageFormat
  let width: Int
  let height: Int
  let byteCount: Int64
  let sourceByteCount: Int64
}

enum AttachmentSessionStatus: String {
  case idle
  case staged
  case prepared
  case uploaded
}

final class AttachmentSession {
  let target: AttachmentTargetIdentity
  var stagingFileName: String?
  var prepared: AttachmentPreparedImage?
  var remotePath: String?
  var lastErrorCode = ""
  var lastMessage = ""

  init(target: AttachmentTargetIdentity) { self.target = target }

  var status: AttachmentSessionStatus {
    if prepared != nil, remotePath != nil { return .uploaded }
    if prepared != nil { return .prepared }
    if stagingFileName != nil { return .staged }
    return .idle
  }

  func recordError(_ errorCode: String, _ message: String) {
    lastErrorCode = errorCode
    lastMessage = message
  }

  func clearError() {
    lastErrorCode = ""
    lastMessage = ""
  }

  func snapshot(previewUri: String) -> [String: Any] {
    [
      "status": status.rawValue,
      "fileId": prepared?.fileName ?? "",
      "previewUri": prepared != nil ? previewUri : "",
      "format": prepared?.format.rawValue ?? "",
      "width": prepared?.width ?? 0,
      "height": prepared?.height ?? 0,
      "byteCount": prepared?.byteCount ?? 0,
      "sourceByteCount": prepared?.sourceByteCount ?? 0,
      "remotePath": remotePath ?? "",
      "errorCode": lastErrorCode,
      "message": lastMessage,
    ]
  }
}

/// Result-dictionary builders matching Attachment*.types.ts one-for-one.
enum AttachmentResults {
  static func held(_ reason: String) -> [String: Any] {
    ["status": "held", "reason": reason]
  }

  static func unavailable(_ reason: String) -> [String: Any] {
    ["status": "unavailable", "reason": reason]
  }

  static func error(_ errorCode: String, _ message: String) -> [String: Any] {
    ["status": "error", "errorCode": errorCode, "message": message]
  }

  static func picked(token: String, byteCount: Int64) -> [String: Any] {
    ["status": "picked", "token": token, "byteCount": byteCount]
  }

  static func canceled() -> [String: Any] { ["status": "canceled"] }

  static func prepared(_ image: AttachmentPreparedImage, previewUri: String) -> [String: Any] {
    [
      "status": "prepared",
      "fileId": image.fileName,
      "previewUri": previewUri,
      "format": image.format.rawValue,
      "width": image.width,
      "height": image.height,
      "byteCount": image.byteCount,
      "sourceByteCount": image.sourceByteCount,
    ]
  }

  static func uploaded(_ remotePath: String) -> [String: Any] {
    ["status": "uploaded", "remotePath": remotePath]
  }

  static func deleted() -> [String: Any] { ["status": "deleted"] }

  static func inserted() -> [String: Any] { ["status": "inserted"] }
}

File: modules/meeterm-terminal/android/src/main/java/dev/meeterm/terminal/AttachmentSession.kt
package dev.meeterm.terminal

/**
 * Issue #28 one-attachment session model.
 *
 * JS sees only narrow result maps; file paths stay in the staging/prepared
 * name fields, and [snapshot] returns display metadata plus the app-local
 * preview URI for the pending image.
 *
 * [AttachmentTargetIdentity] is pane-scoped identity for display and
 * session binding. The Rust core resolves the owning SSH endpoint, runtime,
 * and remote pane itself from the pane's native terminal id
 * (`meeterm_attachment_intent`), so the adapter never records or reuses an
 * SSH owner id.
 */
internal data class AttachmentTargetIdentity(
  val terminalId: String,
  val paneId: String,
  val workspaceId: String,
) {
  companion object {
    fun fromMap(raw: Map<String, Any?>?): AttachmentTargetIdentity? {
      if (raw == null) return null
      val terminalId = raw["terminalId"] as? String ?: return null
      val paneId = raw["paneId"] as? String ?: return null
      val workspaceId = raw["workspaceId"] as? String ?: return null
      return AttachmentTargetIdentity(
        terminalId = terminalId,
        paneId = paneId,
        workspaceId = workspaceId,
      )
    }
  }
}

internal data class AttachmentPreparedImage(
  val fileName: String,
  val format: AttachmentImageFormat,
  val width: Int,
  val height: Int,
  val byteCount: Long,
  val sourceByteCount: Long,
)

internal enum class AttachmentSessionStatus { IDLE, STAGED, PREPARED }

internal data class AttachmentSession(
  val target: AttachmentTargetIdentity,
  var stagingFileName: String? = null,
  var prepared: AttachmentPreparedImage? = null,
  var lastErrorCode: String = "",
  var lastMessage: String = "",
  /**
   * Live `meeterm_attachment_intent` id for the recorded pane terminal; 0
   * means no intent is held (core contract pending or the pane refused).
   * Kept for the process lifetime across sheet unmounts; released only by
   * Discard or a fresh `begin`.
   */
  var intentId: Long = 0,
) {
  /** Live Rust-owned operation; its snapshot stays authoritative. */
  val machine = AttachmentOpMachine()

  val status: AttachmentSessionStatus
    get() = when {
      prepared != null -> AttachmentSessionStatus.PREPARED
      stagingFileName != null -> AttachmentSessionStatus.STAGED
      else -> AttachmentSessionStatus.IDLE
    }

  fun recordError(errorCode: String, message: String) {
    lastErrorCode = errorCode
    lastMessage = message
  }

  fun clearError() {
    lastErrorCode = ""
    lastMessage = ""
  }

  fun snapshot(previewUri: String): Map<String, Any?> = mapOf(
    "status" to status.name.lowercase(),
    "fileId" to (prepared?.fileName ?: ""),
    "previewUri" to if (prepared != null) previewUri else "",
    "format" to (prepared?.format?.name?.lowercase() ?: ""),
    "width" to (prepared?.width ?: 0),
    "height" to (prepared?.height ?: 0),
    "byteCount" to (prepared?.byteCount ?: 0L),
    "sourceByteCount" to (prepared?.sourceByteCount ?: 0L),
    "target" to mapOf(
      "terminalId" to target.terminalId,
      "paneId" to target.paneId,
      "workspaceId" to target.workspaceId,
    ),
    "operation" to machine.operation?.let(AttachmentResults::operation),
    "errorCode" to lastErrorCode,
    "message" to lastMessage,
  )
}

/** Result-map builders matching Attachment*.types.ts one-for-one. */
internal object AttachmentResults {
  fun held(reason: String): Map<String, Any?> = mapOf(
    "status" to "held",
    "reason" to reason,
  )

  fun unavailable(reason: String): Map<String, Any?> = mapOf(
    "status" to "unavailable",
    "reason" to reason,
  )

  fun error(errorCode: String, message: String): Map<String, Any?> = mapOf(
    "status" to "error",
    "errorCode" to errorCode,
    "message" to message,
  )

  fun picked(token: String, byteCount: Long): Map<String, Any?> = mapOf(
    "status" to "picked",
    "token" to token,
    "byteCount" to byteCount,
  )

  fun canceled(): Map<String, Any?> = mapOf("status" to "canceled")

  fun prepared(
    image: AttachmentPreparedImage,
    previewUri: String,
  ): Map<String, Any?> = mapOf(
    "status" to "prepared",
    "fileId" to image.fileName,
    "previewUri" to previewUri,
    "format" to image.format.name.lowercase(),
    "width" to image.width,
    "height" to image.height,
    "byteCount" to image.byteCount,
    "sourceByteCount" to image.sourceByteCount,
  )

  /** Uniform accepted answer carrying the decimal u64 attachment id. */
  fun accepted(attachmentId: Long): Map<String, Any?> = mapOf(
    "status" to "accepted",
    "attachmentId" to java.lang.Long.toUnsignedString(attachmentId),
  )

  /** Core snapshot fields, mirroring AttachmentOperationSnapshot (TS). */
  fun operation(op: AttachmentOperation): Map<String, Any?> = mapOf(
    "phase" to op.wirePhase,
    "attachmentId" to java.lang.Long.toUnsignedString(op.attachmentId),
    "bytesUploaded" to op.bytesUploaded.toDouble(),
    "sizeBytes" to op.sizeBytes.toDouble(),
    "remotePath" to op.remotePath,
    "displayName" to op.displayName,
    "errorCode" to op.errorCode,
    "errorMessage" to op.errorMessage,
    "insertUnconfirmed" to op.insertUnconfirmed,
    "remoteRemoved" to op.remoteRemoved,
    "jobInFlight" to op.jobInFlight,
  )

  fun snapshotResult(op: AttachmentOperation?): Map<String, Any?> =
    if (op == null) {
      mapOf("status" to "idle")
    } else {
      mapOf("status" to "snapshot", "operation" to operation(op))
    }
}
